File: address/change_ofm.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LINE_LENGTH 100
#define OFFSET 2916        // Khoảng cách dòng
#define MAX_PE 16          // Tổng số PE: từ PE0 đến PE15

// Đếm tổng số dòng trong file
int count_lines(const char *filename) {
    FILE *file = fopen(filename, "r");
    if (!file) return -1;

    int count = 0;
    char line[MAX_LINE_LENGTH];
    while (fgets(line, sizeof(line), file)) {
        count++;
    }

    fclose(file);
    return count;
}

// Hàm xử lý 1 PE
void process_pe_file(int pe_id) {
    char inputFile[256];
    char outputFile[256];

    // Tạo tên file input/output theo PE
    sprintf(inputFile, "address/OFM_PE%d.hex", pe_id);
    sprintf(outputFile, "address/OFM_PE%d_change.hex", pe_id);

    // Đếm số dòng trong file
    int total_lines = count_lines(inputFile);
    if (total_lines < OFFSET + 1) {
        printf("⚠️ PE%d: File không đủ %d + 1 dòng để ghép!\n", pe_id, OFFSET);
        return;
    }

    // Mở file input
    FILE *inFile = fopen(inputFile, "r");
    if (!inFile) {
        printf("❌ PE%d: Không thể mở file input '%s'\n", pe_id, inputFile);
        return;
    }

    // Cấp phát bộ nhớ mảng dòng
    char **lines = (char **)malloc(sizeof(char *) * total_lines);
    if (!lines) {
        printf("❌ PE%d: Lỗi cấp phát bộ nhớ!\n", pe_id);
        fclose(inFile);
        return;
    }

    for (int i = 0; i < total_lines; i++) {
        lines[i] = (char *)malloc(MAX_LINE_LENGTH);
        fgets(lines[i], MAX_LINE_LENGTH, inFile);
    }
    fclose(inFile);

    // Mở file output để ghi
    FILE *outFile = fopen(outputFile, "w");
    if (!outFile) {
        printf("❌ PE%d: Không thể mở file output '%s'\n", pe_id, outputFile);
        for (int i = 0; i < total_lines; i++) free(lines[i]);
        free(lines);
        return;
    }

    // Ghi các dòng theo thứ tự: dòng i → dòng i + OFFSET
    for (int i = 0; i < total_lines - OFFSET; i++) {
        fputs(lines[i], outFile);
        fputs(lines[i + OFFSET], outFile);
    }

    printf("✅ PE%d: Đã ghi xong vào '%s'\n", pe_id, outputFile);

    // Giải phóng bộ nhớ
    for (int i = 0; i < total_lines; i++) free(lines[i]);
    free(lines);
    fclose(outFile);
}

int main() {
    for (int pe = 0; pe < MAX_PE; pe++) {
        process_pe_file(pe);
    }

    printf("\n🚀 Đã xử lý xong tất cả %d PE!\n", MAX_PE);
    return 0;
}

File: address/change_ofm.py
import os
import argparse
MAX_LINE_LENGTH = 100

# Đếm tổng số dòng trong file
def count_lines(filename):
    try:
        with open(filename, 'r') as file:
            return sum(1 for line in file)
    except FileNotFoundError:
        print(f"❌ File '{filename}' không tồn tại!")
        return -1

# Hàm xử lý 1 PE
def process_pe_file(pe_id, OFFSET, tile):
    input_file = f"../Fused-Block-CNN/address/OFM_PE{pe_id}.hex"
    output_file = f"../Fused-Block-CNN/address/OFM_PE{pe_id}_change.hex"

    # Đếm số dòng trong file
    total_lines = count_lines(input_file)
    if total_lines < OFFSET + 1:
        print(f"⚠️ PE{pe_id}: File không đủ {OFFSET + 1} dòng để ghép!")
        return

    try:
        with open(input_file, 'r') as in_file:
            lines = in_file.readlines()

        # Mở file output để ghi
        with open(output_file, 'w') as out_file:
            # Ghi các dòng theo thứ tự: dòng i → dòng i + OFFSET → dòng i + a * OFFSET (a từ 1 đến 8)
            for i in range(total_lines - OFFSET):
                # Chuyển thành chữ thường
                lines[i] = lines[i].lower()
                lines[i + OFFSET] = lines[i + OFFSET].lower()

                # Ghi dòng i và dòng i + OFFSET vào file output
                out_file.write(lines[i])
                out_file.write(lines[i + OFFSET])

                # Ghi thêm các dòng i + a * OFFSET, với a từ 2 đến 8
                for a in range(2, tile + 1):
                    index = i + a * OFFSET
                    if index < total_lines:
                        lines[index] = lines[index].lower()
                        out_file.write(lines[index])

        print(f"✅ PE{pe_id}: Đã ghi xong vào '{output_file}'")

    except FileNotFoundError:
        print(f"❌ PE{pe_id}: Không thể mở file input '{input_file}' hoặc file output '{output_file}'.")

# Hàm chính

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--ofm_width", type=int, required=True)
    parser.add_argument("--max_pe", type=int, default=16)
    parser.add_argument("--weight_filter", type = int, required=True)
    args = parser.parse_args()
    tile = args.weight_filter // args.max_pe
    OFFSET = args.ofm_width * args.ofm_width   # Khoảng cách dòng
    for pe in range(args.max_pe):
        process_pe_file(pe, OFFSET, tile)

    print(f"\n🚀 Đã xử lý xong tất cả {args.max_pe} PE!")

if __name__ == "__main__":
    main()